src/field_theory/separable_degree.lean

```diff
@@ -0,0 +1,81 @@
+/-
+Copyright (c) 2021 Jakob Scholbach. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jakob Scholbach
+-/
+import field_theory.separable
+import algebra.algebra.basic
+import data.polynomial.degree
+import algebra.char_p.exp_char
+/-!
+# Separable degree
+
+This file contains basics about the separable degree of a polynomial.
+
+## Main results
+- `separable_degree`: the definition of the separable degree
+- `irreducible_has_separable_contraction`: any nonzero irreducible polynomial can be contracted
+  to a separable polynomial
+- `separable_degree_dvd_degree'`: the separable degree divides the degree, in function of the
+  exponential characteristic of the field
+- `separable_degree_dvd_degree` and `separable_degree_eq_degree` specialize the statement of
+  `separable_degree_dvd_degree`
+
+## Tags
+separable degree, degree, polynomial
+-/
+
+noncomputable theory
+open_locale classical
+
+variables {F : Type} [comm_semiring F]
+  (q : ℕ) [exp_char F q]
+
+/-- A separable contraction of a polynomial `f` is a separable polynomial `g` such that
+`g(x^(q^m)) = f(x)` for some `m : ℕ`.-/
+class has_separable_contraction (f : polynomial F) : Prop :=
+(has_separable_contraction' : ∃ g : polynomial F,
+  g.separable ∧ ∃ m : ℕ, polynomial.expand F (q^m) g = f)
+
+/-- A choice of a separable contraction. -/
+def separable_contraction (f : polynomial F) [hf : has_separable_contraction q f] :
+  polynomial F := classical.some hf.has_separable_contraction'
+
+/-- The separable degree of a polynomial is the degree of a given separable contraction. -/
+def separable_degree (f : polynomial F) [has_separable_contraction q f] : ℕ :=
```

Comment:
Could you please include a proof that this does not depend on the chosen separable contration?

Reply:
Done, now at `theorem separable_degree_eq {n n' : ℕ} :`.